rust/primary/src/header_waiter.rs

```diff
@@ -0,0 +1,292 @@
+                            for missing in missing {
+                                self.parent_requests.entry(missing.clone()).or_insert_with(|| {
+                                    requires_sync.push(missing);
+                                    (round, now)
+                                });
+                            }
+                            if !requires_sync.is_empty() {
+                                let address = self.committee
+                                    .primary(&author)
+                                    .expect("Author of valid header not in the committee")
+                                    .primary_to_primary;
+                                let message = PrimaryMessage::CertificatesRequest(requires_sync, self.name);
+                                let bytes = bincode::serialize(&message).expect("Failed to serialize cert request");
+                                self.network.send(address, Bytes::from(bytes)).await;
+                            }
+                        }
+                    }
+                },
+
+                Some(result) = waiting.next() => match result {
+                    Ok(Some(header)) => {
+                        let _ = self.pending.remove(&header.id);
+                        for x in header.payload.keys() {
+                            let _ = self.batch_requests.remove(x);
+                        }
+                        for x in &header.parents {
+                            let _ = self.parent_requests.remove(x);
+                        }
+                        self.tx_core.send(header).await.expect("Failed to send header");
+                    },
+                    Ok(None) => {
+                        // This request has been canceled.
+                    },
+                    Err(e) => {
+                        error!("{}", e);
+                        panic!("Storage failure: killing node.");
+                    }
+                },
+
+                () = &mut timer => {
+                    // We optimistically sent sync requests to a single node. If this timer triggers,
+                    // it means we were wrong to trust it. We are done waiting for a reply and we now
+                    // broadcast the request to all nodes.
+                    let now = SystemTime::now()
+                        .duration_since(UNIX_EPOCH)
+                        .expect("Failed to measure time")
+                        .as_millis();
+
+                    let mut retry = Vec::new();
+                    for (digest, (_, timestamp)) in &self.parent_requests {
+                        if timestamp + (self.sync_retry_delay as u128) < now {
+                            debug!("Requesting sync for certificate {} (retry)", digest);
+                            retry.push(digest.clone());
+                        }
+                    }
+
+                    let addresses = self.committee
```

Comment:
Should we sync slower than 1 or all? It can become a scalability bottleneck

Reply:
I agree, hopefully issue #10 capture that as well